### DISCUSSION TITLE
API-only flag on Collections and Fields

### DISCUSSION THREAD

**Discussion by ghost** [directus/directus#3611] [Feature Requests]
System collections and their default fields prohibits changes by Users. New fields can be added to capture additional information, e.g. tags, without affecting its core functionalities.

It would be nice to have the similar preventative mechanism applied to application-core schema. As title suggested, an `api_only` attribute can be added to collection and field meta. This field is rendered as read-only in the App.

If it is set on a collection, no changes to the collection and its fields can be made via the App UI. This would supersed the field-level attribute of the same collection.

If it is set on a field, no changes can be made to the field and its collection is protected from being deleted. New fields can still be added/updated/deleted for this collection.

This is a useful feature to prevent admin mistakes from breaking integration with the front-end and other (micro) services.

**Comment by rijkvanzanten** [comment]
To me, this would still be "dangerous", as the admin can still do everything through the API 🤔 

It sounds like we might want to extend this to be a "lock" on/off that locks a field from being edited period

**Reply by benhaynes** [reply]
This is what I had recommended on Discord. I also mentioned this was similar to our _schema lock_ idea.

**Reply by ghost** [reply]
Agreed from the product point of view, and it should be the end goal. The idea works for our use cases. Let me try again.

Instead of `api-only` flag, it would be called `protected`. Combining it with a production flag (unsure how this can be set/toggled), the protective mechanism is only effective during `production`.

**Comment by AustinPhillipTaylor** [comment]
@rijkvanzanten @benhaynes Curious if there has been any plan or progress on this? I'm working on a module whereas the associated collections will require a very specific schema (The module will generate them upon installation/first open). It'll be nice to have a UI lock, so to speak, so administrators can't go in and accidentally remove or alter critical fields. Maybe even have a label in the data model that reads "Managed by [extension_name]".

**Reply by rijkvanzanten** [reply]
Not as of right now. That being said, your specific use ("how to include schema with custom extensions") sounds like a good discussion to have outside of this particular "prevent editing collections" discussion 🙂
